src/Psalm/Internal/Analyzer/Statements/Expression/SimpleTypeInferer.php

```diff
@@ -547,4 +413,200 @@ public static function infer(
+                return null;
+            }
+        }
+
+        $item_key_type = null;
+        if ($array_creation_info->item_key_atomic_types) {
+            $item_key_type = TypeCombiner::combine(
+                $array_creation_info->item_key_atomic_types,
+                null,
+                false,
+                true,
+                30
+            );
+        }
+
+        $item_value_type = null;
+        if ($array_creation_info->item_value_atomic_types) {
+            $item_value_type = TypeCombiner::combine(
+                $array_creation_info->item_value_atomic_types,
+                null,
+                false,
+                true,
+                30
+            );
+        }
+
+        // if this array looks like an object-like array, let's return that instead
+        if ($item_value_type
+            && $item_key_type
+            && ($item_key_type->hasString() || $item_key_type->hasInt())
+            && $array_creation_info->can_create_objectlike
+            && $array_creation_info->property_types
+        ) {
+            $objectlike = new Type\Atomic\TKeyedArray(
+                $array_creation_info->property_types,
+                $array_creation_info->class_strings
+            );
+            $objectlike->sealed = true;
+            $objectlike->is_list = $array_creation_info->all_list;
+            return new Type\Union([$objectlike]);
+        }
+
+        if (!$item_key_type || !$item_value_type) {
+            return null;
+        }
+
+        return new Type\Union([
+            new Type\Atomic\TNonEmptyArray([
+                $item_key_type,
+                $item_value_type,
+            ]),
+        ]);
+    }
+
+    /**
+     * @param   ?array<string, ClassConstantStorage> $existing_class_constants
+     */
+    private static function handleArrayItem(
+        \Psalm\Codebase $codebase,
+        \Psalm\Internal\Provider\NodeDataProvider $nodes,
+        ArrayCreationInfo $array_creation_info,
+        int $int_offset,
+        PhpParser\Node\Expr\ArrayItem $item,
+        \Psalm\Aliases $aliases,
+        \Psalm\FileSource $file_source = null,
+        ?array $existing_class_constants = null,
+        ?string $fq_classlike_name = null
+    ): bool {
+        $single_item_key_type = null;
+
+        if ($item->key) {
+            $single_item_key_type = self::infer(
+                $codebase,
+                $nodes,
+                $item->key,
+                $aliases,
+                $file_source,
+                $existing_class_constants,
+                $fq_classlike_name
+            );
+
+            if ($single_item_key_type) {
+                $array_creation_info->item_key_atomic_types = array_merge(
+                    $array_creation_info->item_key_atomic_types,
+                    array_values($single_item_key_type->getAtomicTypes())
+                );
+            }
+        } else {
+            $array_creation_info->item_key_atomic_types[] =  new Type\Atomic\TInt();
```

Comment:
We could probably directly put a TLiteralInt with $int_offset in here I think

Reply:
Perhaps. That wouldn't be a refactoring in its strict sense though :smile:

Reply:
fair enough :)